### PR TITLE
fix: don't allow blobs to be deleted if in use

### PR DESCRIPTION
**Pull request by peusebiu**
**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/api/routes.go [hunk 1]
```diff
@@ -1052,6 +1052,10 @@ func (rh *RouteHandler) DeleteBlob(response http.ResponseWriter, request *http.R
 			zcommon.WriteJSON(response,
 				http.StatusNotFound,
 				apiErr.NewErrorList(apiErr.NewError(apiErr.BLOB_UNKNOWN, map[string]string{".String()": digest.String()})))
+		} else if errors.Is(err, zerr.ErrBlobBusy) {
+			zcommon.WriteJSON(response,
+				http.StatusConflict,
+				apiErr.NewErrorList(apiErr.NewError(apiErr.BLOB_UNKNOWN, map[string]string{".String()": digest.String()})))
```
Comment (andaaron) [inline pkg/api/routes.go:1058]: Add a new error under pkg/api/errors?

File: pkg/storage/common/common.go [hunk 2]
```diff
@@ -457,6 +457,81 @@ func PruneImageManifestsFromIndex(imgStore storageTypes.ImageStore, repo string,
...
+
+		if found {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func BlobInUse(imgStore storageTypes.ImageStore, repo string,
```
Comment (andaaron) [inline pkg/storage/common/common.go:519]: Function name should contain a verb.
Same applies for the other functions defined in this file.

**Review by andaaron — COMMENTED** [review]
What is the blob is in 'use' meaning it is tagged? Can we delete it? What is the spec saying about this case?

File: pkg/api/routes.go (hunk 1 above)
Comment (peusebiu) [reply pkg/api/routes.go:1058]: can't do it, api errors should be compliant with dist-spec.

**Comment by peusebiu** [comment]
> What is the blob is in 'use' meaning it is tagged? Can we delete it? What is the spec saying about this case?

they talked about this here:
https://github.com/opencontainers/distribution-spec/issues/134

File: pkg/api/routes.go (hunk 1 above)
Comment (andaaron) [reply pkg/api/routes.go:1060]: I think we should use 405 Method Not Allowed, in the spec they mention 405 as a valid response for failure to delete.
I think we should use the `DENIED` from the dist-spec.

```

```

File: pkg/api/routes.go [hunk 3]
```diff
@@ -1054,6 +1054,10 @@ func (rh *RouteHandler) DeleteBlob(response http.ResponseWriter, request *http.R
 			zcommon.WriteJSON(response,
 				http.StatusNotFound,
 				apiErr.NewErrorList(apiErr.NewError(apiErr.BLOB_UNKNOWN, map[string]string{".String()": digest.String()})))
+		} else if errors.Is(err, zerr.ErrBlobReferenced) {
+			zcommon.WriteJSON(response,
+				http.StatusConflict,
```
Comment (andaaron) [inline pkg/api/routes.go:1059]: This is not a valid response code per spec.

**Comment by rchincha** [comment]
https://github.com/project-zot/zot/pull/1526

File: pkg/api/routes.go (hunk 3 above)
Comment (rchincha) [reply pkg/api/routes.go:1059]: 405 status code?

Comment (andaaron) [reply pkg/api/routes.go:1059]: yes

**Review by andaaron — DISMISSED** [review]
This is good from my side. Let's wait for approval from @rchincha

File: examples/config-conformance.json [hunk 4]
```diff
@@ -2,8 +2,8 @@
   "distSpecVersion": "1.1.0-dev",
   "storage": {
     "rootDirectory": "/tmp/zot",
-    "gc": false,
-    "dedupe": false
+    "gc": true,
```
Comment (rchincha) [inline examples/config-conformance.json:5]: Revert this file. I will change this in a separate PR.

File: .github/workflows/oci-conformance-action.yml [hunk 5]
```diff
@@ -34,8 +34,16 @@ jobs:
           RUNNER_TRACKING_ID="" && ./bin/zot-linux-amd64 serve examples/config-conformance.json &
           IP=`hostname -I | awk '{print $1}'`
           echo "SERVER_URL=http://${IP}:8080" >> $GITHUB_ENV
-    - name: Run OCI Distribution Spec conformance tests
-      uses: opencontainers/distribution-spec@main
+    - uses: actions/checkout@v3
+      with:
+        repository: opencontainers/distribution-spec
```
Comment (rchincha) [inline .github/workflows/oci-conformance-action.yml:39]: Point to Brandon's tree like in Josh's PR.

Comment (rchincha) [reply .github/workflows/oci-conformance-action.yml:39]: I will fix in my next PR.

**Review by andaaron — APPROVED** [review]

**Pull request merged**